src/state.rs

```diff
@@ -67,17 +70,36 @@ impl State {
         Ok(())
     }
 
-    #[allow(dead_code)]
-    pub fn register_input(&self, port: u16) -> Result<()> {
+    pub fn register_input(&self, request: RegisterInputRequest) -> Result<()> {
+        let RegisterInputRequest { id, port } = request;
         // TODO: add validation if input already relisted
-        self.pipeline.add_input(InputId(port.into()));
+        self.pipeline.add_input(id.clone());
         self.inputs.insert(
-            port,
+            id.clone(),
             Input {
                 port,
-                rtp_receiver: RtpReceiver::new(self.pipeline.clone(), port),
+                rtp_receiver: RtpReceiver::new(self.pipeline.clone(), port, id),
             },
         );
         Ok(())
     }
+
+    pub fn update_scene(&self, scene: SceneSpec) -> Result<()> {
+        scene.validate(
+            &self.inputs.keys().into_iter().map(|i| i.0).collect(),
+            &self.outputs.keys().into_iter().map(|i| i.0).collect(),
+        )?;
+        self.pipeline.update_scene(scene)?;
```

Comment:
looking at this, I feel we could use the typestate pattern with validation. It's not necessary, but it would make this code way more, for lack of a better word, satisfying.

Reply:
I never heard about typestate patterns before, but after googling a bit I think I understand it, but don't really see what are you suggesting here, can you clarify that?

Reply:
Sure can! I mean that the request parser could give us one struct, called maybe `UnvalidatedSceneSpec`. This struct would have a `validate` method that would consume it and return a different struct, e.g. `ValidatedSceneSpec` (probably with all of the fields of the original one). Only `ValidatedSceneSpec` can be passed to `Pipeline::update_scene`. This would force us to do the validation using the type system. As I said, this is by no means necessary, but, at least for me, feels super satisfying and like a good thing to do.